### ISSUE TITLE
Construct a Block Immediately After An Upgrade

### ISSUE THREAD

**Issue by xcthulhu** [casper-network/casper-node#1666]
- [x] Simplify `EffectBuilder::get_era_validators`
- [x] Simplify fetching global storage in `run_fast_sync_task`
- [X] Reactivate test as per https://github.com/casper-network/sre/issues/19
- [X] Reactivate emergency restart nctl tests
       To paraphrase @fizyk20:
        --------------------------
        Run `nctl-compile && nctl-assets-setup && nctl-start`, and then 
        ```bash
        . utils/nctl/sh/scenarios/emergency_upgrade_test.sh
       ```
       That's one, the second one works the same way, just the name of the script is different (`emergency_upgrade_test_balances.sh`, IIRC).
        You can also try just running the script instead of sourcing it, but there have been some issues with that - sourcing is safer.
       --------------------------
   - [X] ~Get the test to pass on `feat-fast-sync`
         If it doesn't work at first, consider cherry-picking the following commit into `feat-fast-sync` d21b954dad6cd6294dca063696647e03e0ed30dd ; might require additional changes~
   - [X] ~Get the test to pass on https://github.com/casper-network/casper-node/pull/1796~

(Restoring emergency restarts for fast-sync is now ticket #1665)

**Comment by afck** [comment]
So far, I only just got `emergency_upgrade_test.sh` to run at all. Thanks, @fizyk20 for helping me resolve these issues:
* nctl compiles the node with the `casper-mainnet` feature but as a dependency of the client it does _not_ have the feature.
* The `mem_pool_prune_interval` config option is in some of the config files but not in the code. (Maybe something went wrong when rebasing?)

With this change the test starts:

```diff
diff --git a/Makefile b/Makefile
index bccf076f1..89f151ae5 100644
--- a/Makefile
+++ b/Makefile
@@ -9,7 +9,7 @@ CARGO_OPTS := --locked
 CARGO := $(CARGO) $(CARGO_TOOLCHAIN) $(CARGO_OPTS)
 
 DISABLE_LOGGING = RUST_LOG=MatchesNothing
-LEGACY = RUSTFLAGS='--cfg feature="casper-mainnet"'
+LEGACY = RUSTFLAGS=''
 
 # Rust Contracts
 # Directory names should match crate names
diff --git a/resources/local/config.toml b/resources/local/config.toml
index cbdf298d2..2ee4fe638 100644
--- a/resources/local/config.toml
+++ b/resources/local/config.toml
@@ -235,7 +235,7 @@ enable_mem_deduplication = false
 # Memory duplication garbage collection.
 #
 # Sets the frequency how often the memory pool cache is swept for free references.
-mem_pool_prune_interval = 1024
+# mem_pool_prune_interval = 1024
 
 
 # ===================================
diff --git a/utils/nctl/sh/assets/compile_node.sh b/utils/nctl/sh/assets/compile_node.sh
index ca9dca1ae..96524cfa1 100644
--- a/utils/nctl/sh/assets/compile_node.sh
+++ b/utils/nctl/sh/assets/compile_node.sh
@@ -13,9 +13,9 @@ source "$NCTL"/sh/utils/main.sh
 pushd "$NCTL_CASPER_HOME" || exit
 
 if [ "$NCTL_COMPILE_TARGET" = "debug" ]; then
-    cargo build --package casper-node --features casper-mainnet
+    cargo build --package casper-node
 else
-    cargo build --release --package casper-node --features casper-mainnet
+    cargo build --release --package casper-node
 fi
 
 popd || exit
diff --git a/utils/nctl/sh/staging/build.sh b/utils/nctl/sh/staging/build.sh
index 352faba12..7b261681c 100644
--- a/utils/nctl/sh/staging/build.sh
+++ b/utils/nctl/sh/staging/build.sh
@@ -50,9 +50,9 @@ function set_stage_binaries()
 
     # Set node binary.
     if [ "$NCTL_COMPILE_TARGET" = "debug" ]; then
-        cargo build --package casper-node --features casper-mainnet
+        cargo build --package casper-node
     else
-        cargo build --release --package casper-node --features casper-mainnet
+        cargo build --release --package casper-node
     fi
 
     # Set client binary.
```

It then fails at the upgrade point:

```
{"timestamp":"Aug 06 17:44:44.413","level":"INFO","fields":{"message":"Synchronizing linear chain from: BlockHash(d7fce013f407a0727200174e3983fd1a668653fafd1e1f7ce1fbb8457488d2c2)"},"target":"casper_node::reactor::joiner"}
{"timestamp":"Aug 06 17:44:44.413","level":"INFO","fields":{"message":"started REST server","address":"0.0.0.0:14103"},"target":"casper_node::components::rest_server::http_server"}
{"timestamp":"Aug 06 17:44:44.414","level":"INFO","fields":{"message":"started event stream server","address":"0.0.0.0:18103"},"target":"casper_node::components::event_stream_server"}
{"timestamp":"Aug 06 17:44:44.414","level":"INFO","fields":{"message":"reactor main loop is ready"},"target":"casper_node::reactor"}
… (lots of connections established, and then:)
{"timestamp":"Aug 06 17:44:46.732","level":"ERROR","fields":{"message":"Could not insert block header: block header parent hash be17..b9a5, post-state hash e8be..1a61, body hash 26bc..3acb, random bit false, accumulated seed a339..127d, timestamp 2021-08-06T15:44:20.992Z, era_end: era_report: era end: slash , reward PubKey::Ed25519(008d..9424): 1808000000000, PubKey::Ed25519(3534..c8da): 1808000000000, PubKey::Ed25519(b309..ce53): 1807999999990, PubKey::Ed25519(d1ff..0f7b): 1808000000000, PubKey::Ed25519(d24b..da3d): 1807999999990 "},"target":"casper_node::components::storage"}
{"timestamp":"Aug 06 17:44:46.732","level":"ERROR","fields":{"message":"fatal error via control announcement","file":"node/src/components/linear_chain_sync.rs","line":"831","msg":"TryingToJoinBeforeLastEmergencyRestartEra { last_emergency_restart_era: EraId(2), trusted_hash: BlockHash(d7fce013f407a0727200174e3983fd1a668653fafd1e1f7ce1fbb8457488d2c2), trusted_block_header: BlockHeader { parent_hash: BlockHash(be17e245c32b5a6bd4d8976073716f04aedad4662bbf0042d8c7ed891ae3b9a5), state_root_hash: e8bea7228f32aad7c6652d9e70b01b83ea89c84c4928fe77dfbf0f5004f11a61, body_hash: 26bc741337db79044cfe4736202bfb64ff70e91e217de8d3c24590d27eb83acb, random_bit: false, accumulated_seed: a339e4ceeb61613a8d038754827c0c0a063b35af53664951bdfb2af0b981127d, era_end: Some(EraEnd { era_report: EraReport { equivocators: [], rewards: {PublicKey::Ed25519(008d4bc33a68556e0524951aa61f98157dcc9ac1ae3e932284aaf1bada129424): 1808000000000, PublicKey::Ed25519(3534d89d9acd25e6239efa80b90e48a125e0378c7701945e2d52b6483012c8da): 1808000000000, PublicKey::Ed25519(b309286de6691b7ff347d357ff28f577310506b8a2e01986cf0eb96475f2ce53): 1807999999990, PublicKey::Ed25519(d1ff08f8a25161ae909a8c20b25f80316acae254be7c62fa764a04129cd60f7b): 1808000000000, PublicKey::Ed25519(d24b029e0be75adfb65d1567f0487b37817d2b1f042265ea603a68767f06da3d): 1807999999990}, inactive_validators: [] }, next_era_validator_weights: {PublicKey::Ed25519(008d4bc33a68556e0524951aa61f98157dcc9ac1ae3e932284aaf1bada129424): 2000000000000008, PublicKey::Ed25519(3534d89d9acd25e6239efa80b90e48a125e0378c7701945e2d52b6483012c8da): 2000000000000010, PublicKey::Ed25519(b309286de6691b7ff347d357ff28f577310506b8a2e01986cf0eb96475f2ce53): 2000000000000002, PublicKey::Ed25519(d1ff08f8a25161ae909a8c20b25f80316acae254be7c62fa764a04129cd60f7b): 2000000000000006, PublicKey::Ed25519(d24b029e0be75adfb65d1567f0487b37817d2b1f042265ea603a68767f06da3d): 2000000000000004} }), timestamp: Timestamp(1628264660992), era_id: EraId(1), height: 21, protocol_version: ProtocolVersion(SemVer { major: 1, minor: 0, patch: 0 }) } }"},"target":"casper_node::reactor"}
```

**Comment by afck** [comment]
We also need to fix the `BlockHeader` display implementation:

```
era_end: era_report: era end: slash , reward PubKey::Ed25519(008d..9424): 1808000000000, PubKey::Ed25519(3534..c8da): 1808000000000, …
```

**Comment by xcthulhu** [comment]
Closed with #1895
